// atom-notify/dist/index.js
// Generated by CoffeeScript 1.10.0
(function() {
  module.exports = function(title, options) {
    var timeOut;
    if (title == null) {
      title = '';
    }
    if (options == null) {
      options = {};
    }
    if (options.timeOut == null) {
      options.timeOut = 5000;
    }
    timeOut = options.timeOut;
    return {
      addInfo: function(message, options) {
        var notification;
        if (options == null) {
          options = {
            dismissable: true
          };
        }
        options.detail = message;
        notification = atom.notifications.addInfo(title, options);
        setTimeout(((function(_this) {
          return function() {
            return notification.dismiss();
          };
        })(this)), timeOut);
        return notification;
      },
      addSuccess: function(message, options) {
        var notification;
        if (options == null) {
          options = {
            dismissable: true
          };
        }
        options.detail = message;
        notification = atom.notifications.addSuccess(title, options);
        setTimeout(((function(_this) {
          return function() {
            return notification.dismiss();
          };
        })(this)), timeOut);
        return notification;
      },
      addError: function(message, options) {
        var notification;
        if (options == null) {
          options = {
            dismissable: true
          };
        }
        options.detail = message;
        notification = atom.notifications.addError(title, options);
        setTimeout(((function(_this) {
          return function() {
            return notification.dismiss();
          };
        })(this)), timeOut);
        return notification;
      },
      addWarning: function(message, options) {
        var notification;
        if (options == null) {
          options = {
            dismissable: true
          };
        }
        options.detail = message;
        notification = atom.notifications.addWarning(title, options);
        setTimeout(((function(_this) {
          return function() {
            return notification.dismiss();
          };
        })(this)), timeOut);
        return notification;
      }
    };
  };

}).call(this);
